specifyweb/specify/tree_ranks.py

```diff
@@ -0,0 +1,345 @@
+    parent_rank_id = parent_rank.rankid if parent_rank_name != 'root' else -1
+    rank_ids = sorted(list(tree_def_item_model.objects.filter(treedef=tree_def).values_list('rankid', flat=True)))
+    parent_rank_idx = rank_ids.index(parent_rank_id) if rank_ids is not None and parent_rank_name != 'root' else -1
+    next_rank_id = rank_ids[parent_rank_idx + 1] if rank_ids is not None and  parent_rank_idx + 1 < len(rank_ids) else None
+    if next_rank_id is None and parent_rank_name != 'root':
+        next_rank_id = maxsize
+
+    # Don't allow rank IDs less than 0, but really shouldn't be less than 2
+    if new_rank_id is not None and next_rank_id is not None and next_rank_id < 0:
+        raise TreeBusinessRuleException("Can't create rank ID less than 0")
+    
+    # Set conditions for rank ID creation
+    is_tree_def_items_empty = rank_ids is None or len(rank_ids) < 1
+    is_new_rank_first = parent_rank_id == -1
+    is_new_rank_last = parent_rank_idx == len(rank_ids) - 1 if rank_ids is not None else True
+    
+    # Set the default ranks and increments depending on the tree type
+    default_tree_ranks, rank_increment = TREE_RANKS_MAPPING.get(tree.lower(), (None, 100))
+
+    # In the future, add this as a function parameter to allow for more flexibility.
+    # use_default_rank_ids can be set to false if you do not want to use the default rank ids.
+    use_default_rank_ids = True
+
+    # Determine if the default rank ID can be used
+    can_use_default_rank_id = (
+        use_default_rank_ids 
+        and default_tree_ranks is not None 
+        and new_rank_name.upper() in default_tree_ranks
+    )
+    
+    # Only use the the default rank id if the fhe following criteria is met: 
+    # - new_rank_name is in the the default ranks set
+    # - the default rank id is not already used
+    # - the default rank is greater than the target rank
+    # - the default rank is less than the current next rank from the target rank
+    if can_use_default_rank_id:
+        default_rank_id = default_tree_ranks[new_rank_name.upper()]
+
+        # Check if the default rank ID is not already used
+        is_default_rank_id_unused = default_rank_id not in rank_ids
+
+        # Check if the default rank ID can be logically placed in the hierarchy
+        is_placement_valid = (
+            is_tree_def_items_empty 
+            or (is_new_rank_first and default_rank_id < next_rank_id) 
+            or (is_new_rank_last and default_rank_id > parent_rank_id)
+            or (default_rank_id > parent_rank_id and default_rank_id < next_rank_id)
+        )
+
+        if is_default_rank_id_unused and is_placement_valid:
+            new_rank_id = default_rank_id
+
+    # Set the new rank id if a default rank id is not available
+    if new_rank_id is None:
+        # If this is the first rank, set the rank id to the default increment
+        if is_tree_def_items_empty:
+            new_rank_id = rank_increment
```

Comment:
So that there will be room to create smaller ranks if needed.  Using a rank ID of zero should not be done unless one is 100% that they will never create a lower rank.